internal/controller/managedcluster_controller.go

```diff
@@ -683,6 +695,357 @@ func (r *ManagedClusterReconciler) machinesAvailable(ctx context.Context, namesp
+		"global": map[string]any{
+			"port":            443,
+			"insecureFlag":    true,
+			"secretName":      secretName,
+			"secretNamespace": ns,
+		},
+		"vcenter": map[string]any{
+			vCl.Spec.Server: map[string]any{
+				"server": vCl.Spec.Server,
+				"datacenters": []string{
+					vMa.Spec.Datacenter,
+				},
+			},
+		},
+		"labels": map[string]any{
+			"region": "k8s-region",
+			"zone":   "k8s-zone",
+		},
+	}
+	ccmCfgYaml, err := yaml.Marshal(ccmCfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to marshal CCM config: %s", err)
+	}
+	cmData := map[string]string{
+		"vsphere.conf": string(ccmCfgYaml),
+	}
+	return makeSecret(secretName, ns, secretData),
+		makeConfigMap("cloud-config", ns, cmData),
+		nil
+}
+
+func generateVSphereCSISecret(vCl *capv.VSphereCluster, vScrt *corev1.Secret, vMa *capv.VSphereMachine) (*corev1.Secret, error) {
+	csiCfg := `
+[Global]
+cluster-id = "{{ .ClusterID }}"
+
+[VirtualCenter "{{ .Vcenter }}"]
+insecure-flag = "true"
+user = "{{ .Username }}"
+password = "{{ .Password }}"
+port = "443"
+datacenters = "{{ .Datacenter }}"
+`
+	type CSIFields struct {
+		ClusterID, Vcenter, Username, Password, Datacenter string
+	}
+
+	fields := CSIFields{
+		ClusterID:  vCl.Name,
+		Vcenter:    vCl.Spec.Server,
+		Username:   string(vScrt.Data["username"]),
+		Password:   string(vScrt.Data["password"]),
+		Datacenter: vMa.Spec.Datacenter,
+	}
+
+	tmpl, err := texttemplate.New("csiCfg").Parse(csiCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate CSI secret (tmpl parse): %s", err)
+	}
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, fields); err != nil {
+		return nil, fmt.Errorf("failed to generate CSI secret (tmpl execute): %s", err)
+	}
+
+	secretData := map[string][]byte{
+		"csi-vsphere.conf": buf.Bytes(),
+	}
+
+	return makeSecret("vcenter-config-secret", "kube-system", secretData), nil
+}
+
+func applyCCMConfigs(ctx context.Context, kubeconfSecret *corev1.Secret, objects ...client.Object) error {
+	clnt, err := makeClientFromSecret(kubeconfSecret)
+	if err != nil {
+		return fmt.Errorf("failed to create k8s client: %s", err)
+	}
+	for _, object := range objects {
+		if err := clnt.Patch(
+			ctx,
+			object,
+			client.Apply,
+			client.FieldOwner("hmc-controller"),
+		); err != nil {
+			return fmt.Errorf("failed to apply CCM config object %s: %s", object.GetName(), err)
+		}
+	}
+	return nil
+}
+
+func makeSecret(name, namespace string, data map[string][]byte) *corev1.Secret {
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: data,
+	}
+	s.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("Secret"))
+	return s
+}
+
+func makeConfigMap(name, namespace string, data map[string]string) *corev1.ConfigMap {
+	c := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Data: data,
+	}
+	c.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ConfigMap"))
```

Comment:
why do we need it? will it mitigate some runtime checks performed by the controller-gen client, the client is not partial-metadata-scoped IINM

same for the `makeSecret` symbol

Reply:
The object will be with empty GVK without it. So you'll have something like:
`invalid object type: /, Kind=`